File: module_rdp/src/main/java/com/xiaoyv/rdp/screen/listener/PinchZoomListener.java
package com.xiaoyv.rdp.screen.listener;

import android.view.ScaleGestureDetector;

import com.xiaoyv.librdp.view.FreeScrollView;
import com.xiaoyv.librdp.view.SessionView;

public class PinchZoomListener extends ScaleGestureDetector.SimpleOnScaleGestureListener {
    private float scaleFactor = 1.0f;
    private final FreeScrollView freeScrollView;
    private final SessionView sessionView;

    public PinchZoomListener(FreeScrollView freeScrollView, SessionView sessionView) {
        this.freeScrollView = freeScrollView;
        this.sessionView = sessionView;
    }

    @Override
    public boolean onScaleBegin(ScaleGestureDetector detector) {
        freeScrollView.setScrollEnabled(false);
        return true;
    }

    @Override
    public boolean onScale(ScaleGestureDetector detector) {
        // 计算比例因子
        scaleFactor *= detector.getScaleFactor();
        scaleFactor = Math.max(SessionView.MIN_SCALE_FACTOR, Math.min(scaleFactor, SessionView.MAX_SCALE_FACTOR));
        sessionView.setZoom(scaleFactor);

        if (!sessionView.isAtMinZoom() && !sessionView.isAtMaxZoom()) {
            // 将滚动原点转换为新的缩放空间
            float transOriginX = freeScrollView.getScrollX() * detector.getScaleFactor();
            float transOriginY = freeScrollView.getScrollY() * detector.getScaleFactor();

            // 将中心点转换为缩放的空间
            float transCenterX = (freeScrollView.getScrollX() + detector.getFocusX()) * detector.getScaleFactor();
            float transCenterY = (freeScrollView.getScrollY() + detector.getFocusY()) * detector.getScaleFactor();

            //滚动变换后的中心/原点的距离与其旧距离之间的差（focusX / Y）
            freeScrollView.scrollBy((int) ((transCenterX - transOriginX) - detector.getFocusX()), (int) ((transCenterY - transOriginY) - detector.getFocusY()));
        }
        return true;
    }

    @Override
    public void onScaleEnd(ScaleGestureDetector de) {
        freeScrollView.setScrollEnabled(true);
    }
}

File: module_rdp/src/main/java/com/xiaoyv/rdp/screen/config/PinchZoomListener.kt
package com.xiaoyv.rdp.screen.config

import android.view.ScaleGestureDetector
import com.blankj.utilcode.util.LogUtils
import com.blankj.utilcode.util.ToastUtils
import com.freerdp.freerdpcore.view.RdpSessionView
import com.xiaoyv.ui.scroll.FreeScrollView
import kotlin.math.max
import kotlin.math.min

/**
 * SessionView 缩放
 */
class PinchZoomListener(
    private val rsvScroll: FreeScrollView,
    private val rsvSession: RdpSessionView
) : ScaleGestureDetector.SimpleOnScaleGestureListener() {
    private var scaleFactor = 1.0f

    override fun onScaleBegin(detector: ScaleGestureDetector): Boolean {
        rsvScroll.setScrollEnabled(false)
        return true
    }

    override fun onScale(detector: ScaleGestureDetector): Boolean {
        // 计算比例因子
        scaleFactor *= detector.scaleFactor
        scaleFactor = max(
            RdpSessionView.MIN_SCALE_FACTOR,
            min(scaleFactor, RdpSessionView.MAX_SCALE_FACTOR)
        )
        rsvSession.zoom = scaleFactor
        if (rsvSession.isNotAtMinZoom && rsvSession.isNotAtMaxZoom) {
            // 将滚动原点转换为新的缩放空间
            val transOriginX: Float = rsvScroll.scrollX * detector.scaleFactor
            val transOriginY: Float = rsvScroll.scrollY * detector.scaleFactor

            // 将中心点转换为缩放空间
            val transCenterX: Float =
                (rsvScroll.scrollX + detector.focusX) * detector.scaleFactor
            val transCenterY: Float =
                (rsvScroll.scrollY + detector.focusY) * detector.scaleFactor

            // 滚动变换中心原点的距离与其旧距离 (focusXY) 之间的差异
            rsvScroll.scrollBy(
                (transCenterX - transOriginX - detector.focusX).toInt(),
                (transCenterY - transOriginY - detector.focusY).toInt()
            )
        }
        return true
    }

    override fun onScaleEnd(de: ScaleGestureDetector) {
        rsvScroll.setScrollEnabled(true)
    }
}
